### ISSUE TITLE
Feature request: cram index

### ISSUE THREAD

**Issue by matthdsm** [jkbonfield/io_lib#49]
Hi,

This might be a little out of scope here, but would you see fit to add an optional cram index output?
Would be nice not to have to run an additional `samtools index` after compression.

Thanks
Matthias

**Comment by jkbonfield** [comment]
I'm not sure I understand this.  io_lib already has a `cram_index` tool which does the same job as `samtools index` on CRAM files.

**Comment by matthdsm** [comment]
Whoops, missed that one! Thanks for the info!

I was aiming at an extra "--create-index" flag, which would automatically generate an index while compressing, but this is also great!

Matthias

**Comment by jkbonfield** [comment]
You're correct that there's no index-on-the-fly option, which would be good.

Htslib/samtools does have this capability.

Out of interest, are you using io_lib as a library or a command line tool (eg scramble)?  If the latter, then I'd probably recommend Samtools these days.  When originally written, io_lib was substantially more performant than htslib and also had a few key bugs fixed (the io_lib pileup interface for example is used in Gap5 because of some major bugs in htslib at the time that weren't getting fixes).  Since then both the performance and robustness/accuracy of htslib has changed considerably, so it makes more sense to use the more tried and trusted package now.

I still update io_lib, but it's more as an experimental thing as being under my control I have free reign to make releases when I need rather than waiting on a sometimes-lengthy release cycle of samtools/htslib.

**Comment by matthdsm** [comment]
Good to hear. We're currently using `scramble` in a couple of production pipelines, since we were assuming it was still superior to `samtools` and it supports some more fine grained compression options.

I know `samtools` has the `--output-fmt-option` option, but I'm not sure the same can be achieved as `scramble`, since the docs are a bit lacking

**Comment by jkbonfield** [comment]
The docs are in a weird place maybe, but the main samtools.1 man page should cover them.

http://www.htslib.org/doc/samtools.html#GLOBAL_COMMAND_OPTIONS covers these.  You can also eg

```
samtools view -O cram,seqs_per_slice=1000,use_bzip2 -o out.cram in.bam
```

If there are scramble options missing in samtools, then please raise an issue on the samtools git repo and we'll look into improving things there.

James

**Comment by matthdsm** [comment]
Thanks! I wasn't aware there were better docs available.
Do you have docs somewhere how the `scramble` presets (e.g. archive) map to the samtools options?

**Comment by jkbonfield** [comment]
Sadly no, other than code!

https://github.com/jkbonfield/io_lib/blob/master/io_lib/cram_io.c#L6392-L6422
https://github.com/samtools/htslib/blob/develop/cram/cram_io.c#L5805-L5838

I wonder if the difference in `seqs_per_slice` for "fast" mode was a typo when I migrated this to htslib!

**Comment by matthdsm** [comment]
Good enough, I can read code and extrapolate from there!

Thanks for all the help!
